File: programmers/STACK_QUEUE/다리를지나는트럭.cpp
#include <string>
#include <vector>
#include <queue>
using namespace std;

typedef struct car{
    //무게
    int _weight;
    //들어온 시각
    int _clock;
}car;

int solution(int bridge_length, int weight, vector<int> truck_weights) {
    int truck_cnt = truck_weights.size();
    int sec = 0;
    int rp = 0;
    int cur_total_weight = 0;
    int completed = 0;
    queue<car> bridge;
    
    while(completed != truck_cnt){
        ++sec;     
        //자동차를 뺀다.       
        if(!bridge.empty() && (sec - bridge.front()._clock == bridge_length))
        {
            cur_total_weight -= bridge.front()._weight;
            bridge.pop();
            ++completed;
            //다 뺏으면
            if(completed == truck_cnt)
            {
                break;
            }
        }   
        //다리에 자동차 넣는다.
        if(truck_weights[rp] + cur_total_weight <= weight){
            bridge.push({truck_weights[rp], sec});
            cur_total_weight += truck_weights[rp];
            ++rp;
        }
    }
    
    return sec;
}

File: programmers/STACK_QUEUE/다리를_지나는_트럭.java
package com.kt.aicare.common.util.programmers.STACK_QUEUE;
import java.util.*;

public class 다리를_지나는_트럭 {
  final static int W = 0;// weight
  final static int T = 1;// time(들어온 시간)

  static class Bridge {
    int totalWeight;
    int weightLimit;
    int length;
    Deque<int[]> queue;

    Bridge(int weightLimit, int length) {
      this.totalWeight = 0;
      this.queue = new LinkedList<>();
      this.weightLimit = weightLimit;
      this.length = length;
    }

    boolean addTruck(int truckWeight, int curTime) {
      if ((totalWeight + truckWeight) > weightLimit) {
        return false;
      }

      this.queue.offerLast(new int[]{truckWeight, curTime});
      this.totalWeight += truckWeight;
      return true;
    }

    boolean removeTruck(int curTime) {
      if (this.queue.isEmpty()) {
        return false;
      }
      int[] cur = this.queue.peek();
      if ((curTime - cur[T]) < length) {
        return false;
      }
      totalWeight -= cur[W];
      this.queue.pollFirst();
      return true;
    }

    boolean isEmpty() {
      return this.queue.isEmpty();
    }

    public String toString() {
      StringBuffer sb = new StringBuffer();
      sb.append("[");
      for (int[] elem : this.queue) {
        sb.append(String.format("(%d, %d), ", elem[W], elem[T]));
      }
      sb.append("]");
      return sb.toString();
    }
  }

  public int solution(int bridge_length, int weight, int[] truck_weights) {
    int timer = 0;

    Deque<Integer> readyQueue = new LinkedList<>();
    Bridge bridge = new Bridge(weight, bridge_length);

    for (int truckWeight : truck_weights) {
      readyQueue.offerLast(truckWeight);
    }

    while(!readyQueue.isEmpty() ||
        !bridge.isEmpty()) {
      bridge.removeTruck(timer);
      if (!readyQueue.isEmpty()) {
        int curWeight = readyQueue.peekFirst();
        if (bridge.addTruck(curWeight, timer)) {
          readyQueue.pollFirst();
        }
      }
      timer++;
    }

    return timer;
  }

}
